lib/dap/filter/geoip2.rb

```diff
@@ -0,0 +1,87 @@
+require 'maxmind/db'
+
+module Dap
+module Filter
+
+require 'dap/utils/misc'
+
+module GeoIP2Library
+  GEOIP_DIRS = [
+    File.expand_path( File.join( File.dirname(__FILE__), "..", "..", "..", "data")),
+    "/var/lib/geoip",
+    "/var/lib/geoip2"
+  ]
+  GEOIP_CITY = %W{ GeoLite2-City.mmdb }
+  GEOIP_ASN = %W{ GeoLite2-ASN.mmdb }
+
+  @@geo_city = nil
+  @@geo_asn = nil
+
+  GEOIP_DIRS.each do |d|
+    GEOIP_CITY.each do |f|
+      path = File.join(d, f)
+      if ::File.exist?(path)
+        @@geo_city = MaxMind::DB.new(path, mode: MaxMind::DB::MODE_MEMORY)
+        break
+      end
+    end
+    GEOIP_ASN.each do |f|
+      path = File.join(d, f)
+      if ::File.exist?(path)
+        @@geo_asn = MaxMind::DB.new(path, mode: MaxMind::DB::MODE_MEMORY)
+        break
+      end
+    end
+  end
+end
+
+
+#
+# Add GeoIP2 tags using the MaxMind GeoIP2::City
+#
+class FilterGeoIP2
+  include BaseDecoder
+  include GeoIP2Library
+  def decode(ip)
+    unless @@geo_city
+      raise "No MaxMind GeoIP2::City data found"
+    end
+    return unless (geo_hash = @@geo_city.get(ip))
+    return Dap::Utils::Misc.flatten_hash(geo_hash)
```

Comment:
we should explicitly set keys as godap does instead of hash merging here. The biggest reason is to insulate the DAP output from any upstream changes to how the hash is constructed (just in case). We do the same thing in godap.

I'm also a fan of keeping the `%s.geoip2.%s` pattern godap has, to isolate all geoip fields with a prefix. This allows easy pruning of all geoip data (which is necessary for cases when you want to use the compat filter -- for this purpose, I created a `remove_prefix` filter). So instead of fields such as `line.city`, we should have`line.city.name`. See https://github.com/rapid7/godap/blob/a426c17c1dbb8fc1846212c206f34d1bd2d1d37a/filter/geoip2.go#L40.

The maxmind database also includes a few localized versions of text data (city names, etc.). For this reason, the name part of the hash is usually a map supporting each locale ("en", "de", "ru", etc.). We should allow configuration of the locale dap uses, but only spit out one to make the output less spammy. Godap is doing this currently with an environment variable `GEOIP2_LANGUAGE`, which defaults to `en`. See https://github.com/rapid7/godap/blob/a426c17c1dbb8fc1846212c206f34d1bd2d1d37a/filter/geoip2.go#L146 and https://github.com/rapid7/godap/blob/a426c17c1dbb8fc1846212c206f34d1bd2d1d37a/filter/geoip2.go#L40.

Reply:
Hmm, I agree on the insulation in some regards but not necessarily on explicitly setting keys; that is a lot of extra code.  Maybe instead here I'll just flatten as I've done and then strip out all but what matches what we like.  

Agreed on the prefix.  Will update.

`dap` has a `match_remove` that does something similar to `remove_prefix`.  IMO we should use that instead and going forward we should consult what `dap` currently has before adding new filters to `godap`.

The language thing makes sense.  I'll see about the best way to handle that here.